src/main/scala/mesosphere/marathon/core/check/Check.scala

```diff
@@ -0,0 +1,245 @@
+      .setIntervalSeconds(this.interval.toSeconds.toInt)
+      .setTimeoutSeconds(this.timeout.toSeconds.toInt)
+      .setDelaySeconds(this.delay.toSeconds.toInt)
+}
+
+sealed trait PortReference extends Product with Serializable {
+  def apply(assignments: Seq[PortAssignment]): PortAssignment
+  def buildProto(builder: Protos.CheckDefinition.Builder): Unit
+}
+
+object PortReference {
+  case class ByIndex(value: Int) extends PortReference {
+    override def apply(assignments: Seq[PortAssignment]): PortAssignment =
+      assignments(value)
+    override def buildProto(builder: Protos.CheckDefinition.Builder): Unit =
+      builder.setPortIndex(value)
+  }
+
+  def apply(value: Int): PortReference = ByIndex(value)
+
+  def fromProto(pb: Protos.CheckDefinition): Option[PortReference] =
+    if (pb.hasPortIndex) Some(ByIndex(pb.getPortIndex))
+    else if (!pb.hasPort) Some(ByIndex(0)) // backward compatibility, this used to be the default value in marathon.proto
+    else None
+}
+
+sealed trait CheckWithPort extends Check {
+  def portIndex: Option[PortReference]
+  def port: Option[Int]
+}
+
+object CheckWithPort {
+  val DefaultPortIndex = None
+  val DefaultPort = None
+
+  import mesosphere.marathon.api.v2.Validation.isTrue
+
+  implicit val Validator: Validator[CheckWithPort] =
+    isTrue("Check must specify either a port or a portIndex") { hc =>
+      hc.portIndex.isDefined ^ hc.port.isDefined
+    }
+}
+
+sealed trait MesosCheck extends Check {
+  def interval: FiniteDuration
+  def timeout: FiniteDuration
+  def delay: FiniteDuration
+
+  override protected def protoBuilder: Protos.CheckDefinition.Builder =
+    super.protoBuilder.setDelaySeconds(delay.toSeconds.toInt)
+
+  def toMesos(portAssignments: Seq[PortAssignment] = Seq.empty): Option[MesosProtos.CheckInfo]
+}
+
+sealed trait MesosCheckWithPorts extends CheckWithPort { this: Check =>
+  def effectivePort(portAssignments: Seq[PortAssignment]): Option[Int] = {
+    port.orElse {
+      val portAssignment: Option[PortAssignment] = portIndex.map(index => index(portAssignments))
+      // Mesos enters the container's network to probe the port, hence we prefer `containerPort`
+      // to `hostPort` here (as opposed to MarathonCheck which is the opposite)
+      portAssignment.flatMap(_.containerPort).orElse(portAssignment.flatMap(_.hostPort))
+    }
+  }
+}
+
+case class MesosCommandCheck(
+    interval: FiniteDuration = Check.DefaultInterval,
+    timeout: FiniteDuration = Check.DefaultTimeout,
+    delay: FiniteDuration = Check.DefaultDelay,
+    command: Executable)
+  extends Check with MesosCheck {
+  override def toProto: Protos.CheckDefinition = {
+    protoBuilder
+      .setProtocol(Protos.CheckDefinition.Protocol.COMMAND)
+      .setCommand(Executable.toProto(command))
+      .build
+  }
+
+  override def toMesos(portAssignments: Seq[PortAssignment] = Seq.empty): Option[MesosProtos.CheckInfo] = {
+    Option(MesosProtos.CheckInfo.newBuilder
```

Comment:
Just so I understand, why are we returning an `Option` here?

Reply:
Right here:

https://github.com/mesosphere/marathon/pull/6824/files/f72d68a64716e86ee587d1bdbbc280c2823d3c1a#diff-523f23b18dfb236b8d87ab530c4a9608R142

Reply:
Oh... I'd already linked to it :) See that link. There is a spot where None can be returned.

Reply:
if we have a HTTP or TCP check we should always have a port which means it shouldn't be an option.
HOWEVER:
1. it is consistent with HealthChecks... do we want to refactor HCs as well?  or be consist with them?
2. if removed, it cascades... eventually to `effectivePort` which would need to be refactored to remove it's Option
3.   One could even claim that the `toMesos` defined in MesosCheck is a poor place to define `toMesos` because the idea of passing in ports doesn't make sense to a MesosCommandCheck.

I'll look a little more... but most of my struggle working on this code includes consistent with HCs since I see checks as the abstract of HCs.

Reply:
yeah... the implementation of `effectivePort` gets muddy without the Option

Reply:
I did remove the default `= Seq.empty`